### ISSUE TITLE
MueLu research and example code build failures for new CUDA ATDM build on hansen/shiller

### ISSUE THREAD

**Issue by bartlettroscoe** [trilinos/Trilinos#2319]
**CC:** @trilinos/muelu, @fryeguy52

## Next Action Status:

Commit eee871d which sets `MueLu_ENABLE_Epertra=OFF` and fixes the build failures.  Next: Determine if `MueLu_ENABLE_Epertra=OFF` is acceptable to EMPIRE ...

## Description

The MueLu package shows build falures for the CUDA ATDM builds today on hansen shown at:

* https://testing.sandia.gov/cdash/index.php?project=Trilinos&filtercount=1&showfilters=1&field1=buildname&compare1=63&value1=-atdm-

for the builds:

* `Trilinos-atdm-hansen-shiller-cuda-debug`: https://testing.sandia.gov/cdash/index.php?project=Trilinos&parentid=3412693
* `Trilinos-atdm-hansen-shiller-cuda-opt`: https://testing.sandia.gov/cdash/index.php?project=Trilinos&parentid=3412702

The build failures for example at:

* https://testing.sandia.gov/cdash/viewBuildError.php?buildid=3412805

all show undefined reference link failues like:

```
CMakeFiles/MueLu_ImportTest.dir/Import.cpp.o: In function `int main_<double, int, int, Kokkos::Compat::KokkosDeviceWrapperNode<Kokkos::Serial, Kokkos::HostSpace> >(Teuchos::CommandLineProcessor&, Xpetra::UnderlyingLib, int, char**)':
/home/jenkins/hansen/workspace/Trilinos-atdm-hansen-shiller-cuda-debug/SRC_AND_BUILD/Trilinos/packages/muelu/research/luc/region_algorithms/Import.cpp:153: undefined reference to `Tpetra::Map<int, int, Kokkos::Compat::KokkosDeviceWrapperNode<Kokkos::Serial, Kokkos::HostSpace> >::Map(unsigned long, Teuchos::ArrayView<int const> const&, int, Teuchos::RCP<Teuchos::Comm<int> const> const&, Teuchos::RCP<Kokkos::Compat::KokkosDeviceWrapperNode<Kokkos::Serial, Kokkos::HostSpace> > const&)'
/home/jenkins/hansen/workspace/Trilinos-atdm-hansen-shiller-cuda-debug/SRC_AND_BUILD/Trilinos/packages/muelu/research/luc/region_algorithms/Import.cpp:154: undefined reference to `Tpetra::Map<int, int, Kokkos::Compat::KokkosDeviceWrapperNode<Kokkos::Serial, Kokkos::HostSpace> >::Map(unsigned long, Teuchos::ArrayView<int const> const&, int, Teuchos::RCP<Teuchos::Comm<int> const> const&, Teuchos::RCP<Kokkos::Compat::KokkosDeviceWrapperNode<Kokkos::Serial, Kokkos::HostSpace> > const&)'
/home/jenkins/hansen/workspace/Trilinos-atdm-hansen-shiller-cuda-debug/SRC_AND_BUILD/Trilinos/packages/muelu/research/luc/region_algorithms/Import.cpp:156: undefined reference to `Tpetra::MultiVector<double, int, int, Kokkos::Compat::KokkosDeviceWrapperNode<Kokkos::Serial, Kokkos::HostSpace> >::MultiVector(Teuchos::RCP<Tpetra::Map<int, int, Kokkos::Compat::KokkosDeviceWrapperNode<Kokkos::Serial, Kokkos::HostSpace> > const> const&, unsigned long, bool)'
/home/jenkins/hansen/workspace/Trilinos-atdm-hansen-shiller-cuda-debug/SRC_AND_BUILD/Trilinos/packages/muelu/research/luc/region_algorithms/Import.cpp:157: undefined reference to `Tpetra::MultiVector<double, int, int, Kokkos::Compat::KokkosDeviceWrapperNode<Kokkos::Serial, Kokkos::HostSpace> >::getDataNonConst(unsigned long)'
/home/jenkins/hansen/workspace/Trilinos-atdm-hansen-shiller-cuda-debug/SRC_AND_BUILD/Trilinos/packages/muelu/research/luc/region_algorithms/Import.cpp:167: undefined reference to `Tpetra::MultiVector<double, int, int, Kokkos::Compat::KokkosDeviceWrapperNode<Kokkos::Serial, Kokkos::HostSpace> >::MultiVector(Teuchos::RCP<Tpetra::Map<int, int, Kokkos::Compat::KokkosDeviceWrapperNode<Kokkos::Serial, Kokkos::HostSpace> > const> const&, unsigned long, bool)'
/home/jenkins/hansen/workspace/Trilinos-atdm-hansen-shiller-cuda-debug/SRC_AND_BUILD/Trilinos/packages/muelu/research/luc/region_algorithms/Import.cpp:169: undefined reference to `Tpetra::Export<int, int, Kokkos::Compat::KokkosDeviceWrapperNode<Kokkos::Serial, Kokkos::HostSpace> >::Export(Teuchos::RCP<Tpetra::Map<int, int, Kokkos::Compat::KokkosDeviceWrapperNode<Kokkos::Serial, Kokkos::HostSpace> > const> const&, Teuchos::RCP<Tpetra::Map<int, int, Kokkos::Compat::KokkosDeviceWrapperNode<Kokkos::Serial, Kokkos::HostSpace> > const> const&)'
/home/jenkins/hansen/workspace/Trilinos-atdm-hansen-shiller-cuda-debug/SRC_AND_BUILD/Trilinos/packages/muelu/research/luc/region_algorithms/Import.cpp:170: undefined reference to `Tpetra::DistObject<double, int, int, Kokkos::Compat::KokkosDeviceWrapperNode<Kokkos::Serial, Kokkos::HostSpace> >::doExport(Tpetra::SrcDistObject const&, Tpetra::Export<int, int, Kokkos::Compat::KokkosDeviceWrapperNode<Kokkos::Serial, Kokkos::HostSpace> > const&, Tpetra::CombineMode)'
/home/jenkins/hansen/workspace/Trilinos-atdm-hansen-shiller-cuda-debug/SRC_AND_BUILD/Trilinos/packages/muelu/research/luc/region_algorithms/Import.cpp:171: undefined reference to `Tpetra::MultiVector<double, int, int, Kokkos::Compat::KokkosDeviceWrapperNode<Kokkos::Serial, Kokkos::HostSpace> >::getData(unsigned long) const'
/home/jenkins/hansen/workspace/Trilinos-atdm-hansen-shiller-cuda-debug/SRC_AND_BUILD/Trilinos/packages/muelu/research/luc/region_algorithms/Import.cpp:169: undefined reference to `Tpetra::Export<int, int, Kokkos::Compat::KokkosDeviceWrapperNode<Kokkos::Serial, Kokkos::HostSpace> >::~Export()'
/home/jenkins/hansen/workspace/Trilinos-atdm-hansen-shiller-cuda-debug/SRC_AND_BUILD/Trilinos/packages/muelu/research/luc/region_algorithms/Import.cpp:167: undefined reference to `Tpetra::MultiVector<double, int, int, Kokkos::Compat::KokkosDeviceWrapperNode<Kokkos::Serial, Kokkos::HostSpace> >::~MultiVector()'
/home/jenkins/hansen/workspace/Trilinos-atdm-hansen-shiller-cuda-debug/SRC_AND_BUILD/Trilinos/packages/muelu/research/luc/region_algorithms/Import.cpp:156: undefined reference to `Tpetra::MultiVector<double, int, int, Kokkos::Compat::KokkosDeviceWrapperNode<Kokkos::Serial, Kokkos::HostSpace> >::~MultiVector()'
/home/jenkins/hansen/workspace/Trilinos-atdm-hansen-shiller-cuda-debug/SRC_AND_BUILD/Trilinos/packages/muelu/research/luc/region_algorithms/Import.cpp:169: undefined reference to `Tpetra::Export<int, int, Kokkos::Compat::KokkosDeviceWrapperNode<Kokkos::Serial, Kokkos::HostSpace> >::~Export()'
/home/jenkins/hansen/workspace/Trilinos-atdm-hansen-shiller-cuda-debug/SRC_AND_BUILD/Trilinos/packages/muelu/research/luc/region_algorithms/Import.cpp:167: undefined reference to `Tpetra::MultiVector<double, int, int, Kokkos::Compat::KokkosDeviceWrapperNode<Kokkos::Serial, Kokkos::HostSpace> >::~MultiVector()'
/home/jenkins/hansen/workspace/Trilinos-atdm-hansen-shiller-cuda-debug/SRC_AND_BUILD/Trilinos/packages/muelu/research/luc/region_algorithms/Import.cpp:156: undefined reference to `Tpetra::MultiVector<double, int, int, Kokkos::Compat::KokkosDeviceWrapperNode<Kokkos::Serial, Kokkos::HostSpace> >::~MultiVector()'
collect2: error: ld returned 1 exit status
```

but each executable has a slightly different set of link failures.

It looks like some explicit template instantiations are missing?

## Steps to Reproduce:

The instructions to reproduce these build failures can be found starting at:

* https://snl-wiki.sandia.gov/display/CoodinatedDevOpsATDM/ATDM+Builds+of+Trilinos

and clicking "Reproducing ATDM builds locally" which takes you to:

* https://github.com/trilinos/Trilinos/blob/develop/cmake/std/atdm/README.md

Basically, on `hansen` or `shiller`, you just clone the Trilinos repo (with location depicted as `$TRILINOS_DIR` below), get on the `develop` branch.  Then create a build directory and do the configure and build as:

```
$ cd <some_build_dir>/

$ source $TRILINOS_DIR/cmake/std/atdm/load-env.sh cuda-opt

$ cmake \
  -DTrilinos_CONFIGURE_OPTIONS_FILE:STRING=cmake/std/atdm/ATDMDevEnv.cmake \
  -DTrilinos_ENABLE_TESTS=ON -DTrilinos_ENABLE_MueLu=ON \
  $TRILINOS_DIR

$ make -j16
```

**Comment by jhux2** [comment]
MueLu's research code  directory should not be enabled by default.  This is likely caused by insufficient cmake guards.

**Comment by mayrmt** [comment]
I checked ```Trilinos/packages/muelu/research/luc/region_algorithms/Import.cpp``` accidentally. We can just revert the commit 506af3b17d4c8a5de1d2bccdf1bd27d859aef1c0 since we don't need this anyways.

**Comment by mayrmt** [comment]
@bartlettroscoe How do I best revert this commit? 
- Reverting it locally and the pushing to Trilinos/develop with the checkin-script?
- Reverting it locally and issuing a pull request?

**Comment by bartlettroscoe** [comment]
> How do I best revert this commit?

Did you push this to 'develop' yet?  If you only committed it locally, then you can remove it in various ways locally as described:

* https://blog.github.com/2015-06-08-how-to-undo-almost-anything-with-git/

Otherwise, send me email and we can converse there.

**Comment by mayrmt** [comment]
PR #2326 should have fixed this. Not closing yet. 

@bartlettroscoe Can you confirm that this is resolved and then close this issue? Thank you!

**Comment by bartlettroscoe** [comment]
> Fix has been merged via PR #2326. Not closing yet to make sure that this actually cured the problem.

@mayrmt,

Given the six link failures that were shown on CDash, I don't think this one PR #2326 will fix all of them but I sure hope I am wrong :-)

> @bartlettroscoe Can you confirm that this is resolved and then close this issue? Thank you!

We will take a look at the ATDM builds dashboard tomorrow and that will be the telling.  I am looking over the those builds every day as we get them cleaned up so if the problem goes away, I will close this issue.

Thanks!

**Comment by bartlettroscoe** [comment]
@mayrmt,

It looks like your new commit 7a945e6 that was pushed on Saturday and pulled on Sunday as shown at:

* https://testing.sandia.gov/cdash/viewNotes.php?buildid=3417530##note5

removed the build error for the file `packages/muelu/research/luc/region_algorithms/Import.cpp` but the other link errors remain as shown in automated testing at:

* https://testing.sandia.gov/cdash/viewBuildError.php?buildid=3417530
* https://testing.sandia.gov/cdash/viewBuildError.php?buildid=3417637

Can someone take a look at these?  I think anyone with access to SRN or SON machines `shiller` or `hansen` can reproduce these failures as described in the "Steps to Reproduce" above.

**Comment by bartlettroscoe** [comment]
Also note that even if you exclude the 8 "Not Run" tests for the missing executables that would not link, there are still 23 failing MueLu tests for these CUDA builds.  I was going to wait until these build failures were fixed before posting new GitHub issues for those failures but some MueLu developer might want to look into those too.  Those can be see at:

* https://testing.sandia.gov/cdash/index.php?project=Trilinos&parentid=3417390
* https://testing.sandia.gov/cdash/index.php?project=Trilinos&parentid=3417509

**Comment by mhoemmen** [comment]
If folks are gone this week at SIAM PP, I can help, just not today so much.

**Comment by jhux2** [comment]
@bartlettroscoe I'm not surprised at these failures, as MueLu has an experimental track CUDA build that hasn't been clean for quite a while.  Could you temporarily disable these for the ATDM build until the MueLu team has time to look at these?  (I am facing a couple March conference/milestone deadlines.)

**Comment by bartlettroscoe** [comment]
> I'm not surprised at these failures, as MueLu has an experimental track CUDA build that hasn't been clean for quite a while. Could you temporarily disable these for the ATDM build until the MueLu team has time to look at these? (I am facing a couple March conference/milestone deadlines.)

@jhux2, okay, I will disable is little as I can to make everything pass.  I will also disable the failing tests as well.  Then someone on the MueLu team can log onto the `shiller` or `hansen` and work out all of these issues when they have time.

**Comment by bartlettroscoe** [comment]
@bathmatt, @jhux2, @mhoemmen, and @srajama1,

I was wrong, I did create a GitHub issue for these MueLu build failures.   Could this be related to the build failures that @bathmatt reported for EMPIRE?  It looks like some explicit instantiations are missing.

**Comment by jhux2** [comment]
> I was wrong, I did create a GitHub issue for these MueLu build failures. Could this be related to the build failures that @bathmatt reported for EMPIRE? It looks like some explicit instantiations are missing.

@bartlettroscoe Yes, in fact the very error @bathmatt reported also appears on the dashboard.

By the way, I see this output during the configure process

```
NOTE: Kokkos::Serial is ON (the CMake option Kokkos_ENABLE_Serial is ON), but the corresponding Tpetra Node type is disabled.  If you want to enable instantiation and use of Kokkos::Serial in Tpetra, please also set the CMake option Tpetra_INST_SERIAL:BOOL=ON.  If you use the Kokkos::Serial Node type in Tpetra without doing this, you will get link errors!
-- Tpetra execution space availability (ON means available): 
--   - Serial:  OFF
--   - Threads: OFF
--   - OpenMP:  OFF
--   - Cuda:    ON
```

The link errors refer to symbols templated on `Kokkos::Compat::KokkosDeviceWrapperNode<Kokkos::Serial, Kokkos::HostSpace>`, so the cmake message might be relevant.

**Comment by bartlettroscoe** [comment]
> By the way, I see this output during the configure process
> 
> ```
> NOTE: Kokkos::Serial is ON (the CMake option Kokkos_ENABLE_Serial is ON), but the corresponding Tpetra Node type is disabled.  If you want to enable instantiation and use of Kokkos::Serial in Tpetra, please also set the CMake option Tpetra_INST_SERIAL:BOOL=ON.  If you use the Kokkos::Serial Node type in Tpetra without doing this, you will get link errors!
> -- Tpetra execution space availability (ON means available): 
> --   - Serial:  OFF
> --   - Threads: OFF
> --   - OpenMP:  OFF
> --   - Cuda:    ON
> ```
> 
> The link errors refer to symbols templated on `Kokkos::Compat::KokkosDeviceWrapperNode<Kokkos::Serial, Kokkos::HostSpace>`, so the cmake message might be relevant.

@trilinos/tpetra developers,

Can we set up Tpetra to automatically enable `Tpetra_INST_SERIAL:BOOL=ON` by default instead of just printing out this warning message?  In general, we need to be looking for ways to make it easier for users to configure Trilinos correctly by just setting obvious options.

Otherwise, I will try enabling this option for the CUDA build on hansen/shiller and see if this makes the link errors go away.

**Comment by jhux2** [comment]
@bartlettroscoe What bothers me is that MueLu should not require that node type.  Enabling `Tpetra_INST_SERIAL:BOOL=ON` might fix this error, but doesn't address the underlying problem.  (The build time and executable size will go up, too.)

**Comment by bartlettroscoe** [comment]
> What bothers me is that MueLu should not require that node type. Enabling Tpetra_INST_SERIAL:BOOL=ON might fix this error, but doesn't address the underlying problem. (The build time and executable size will go up, too.)

@jhux2,

Somehow this is getting manifested in the MueLu test suite itself so one only needs to look at how MueLu is using upstream packages and what those packages are doing.  If you look at the link failure, for example, today at:

* https://testing-vm.sandia.gov/cdash/viewBuildError.php?buildid=3365789

you see undefined reference errors coming directly from MueLu source files as well as from Tpetra files and other packages. 

Can some MueLu developer log into to 'hanen' or 'shiller' and see if they can reproduce these link failures as described above?  It should only take a few minutes of person-time to get the build going.

Otherwise, I will let you know what happens with setting `Tpetra_INST_SERIAL:BOOL=ON` for these CUDA builds.

**Comment by jhux2** [comment]
I'm on shiller now, diagnosing the configure process.  I believe that's where the error is.

**Comment by jhux2** [comment]
@bartlettroscoe I tested two options:
  1. `-DTpetra_ENABLE_Epetra:BOOL=OFF`.  This fixed the error.  However, EMPIRE requires Panzer, which requires Epetra, so this is not a workable option.
  2. `-DTpetra_INST_SERIAL:BOOL=ON`.   This fixed the error, at the cost of increasing the build time and probably any executables.

For now, can you enable option 2?  Longer term, we may be able to relax this requirement in MueLu.

**Comment by mhoemmen** [comment]
@bartlettroscoe wrote:
> Can we set up Tpetra to automatically enable Tpetra_INST_SERIAL:BOOL=ON by default instead of just printing out this warning message? In general, we need to be looking for ways to make it easier for users to configure Trilinos correctly by just setting obvious options.

This used to be ON by default, but that changed as part of Kokkos' refactor of CMake / Makefile options.  Changing it back would increase the build time quite a bit.  Is the problem that downstream packages don't do ETI correctly?  That's an issue very much like https://github.com/trilinos/Trilinos/issues/74 in that what looks like an easy CMake option to set, actually increases the build time a lot and does not help generality.

**Comment by bartlettroscoe** [comment]
> 2. `-DTpetra_INST_SERIAL:BOOL=ON`. This fixed the error, at the cost of increasing the build time and probably any executables.
> 
> For now, can you enable option 2? Longer term, we may be able to relax this requirement in MueLu.

@jhux2,

I also confirmed that setting `Tpetra_INST_SERIAL=ON` fixed the build of the MueLu tests and examples.  I will go ahead and push that updated `ATDMDevEnv.cmake` file.

> This used to be ON by default, but that changed as part of Kokkos' refactor of CMake / Makefile options. Changing it back would increase the build time quite a bit.

@mhoemmen,

So the deal is that as long as Epetra is not enabled then you don't need these instantiations?

Note that we could add some specialized logic to the file `Trilinos/cmake/ProjectCompilerPostConfigure.cmake` that could turn on `Tpetra_INST_SERIAL=ON` when it detects that MueLu Tpetra and Epetra support are enabled.  As shown at:

* https://tribits.org/doc/TribitsDevelopersGuide.html#full-tribits-project-configuration

that file gets processed after the final set of enables and disables are determined and therefore, it would have all of the info needed to determine when this needed to be enabled.  That would save users from having to figure this out on their own.  Again, anything we can do to make Trilinos configure correctly for the requested user configuration will go a long way to reducing the reputation that Trilinos is hard to build (which I hear a lot).

So what are the right set of enable variables to look for in order to set  `Tpetra_INST_SERIAL=ON`?  This will result in this getting enabled only when it needs to be.

**Comment by jhux2** [comment]
@mhoemmen wrote:
> This used to be ON by default, but that changed as part of Kokkos' refactor of CMake / Makefile options. Changing it back would increase the build time quite a bit. Is the problem that downstream packages don't do ETI correctly? That's an issue very much like #74 in that what looks like an easy CMake option to set, actually increases the build time a lot and does not help generality.

Yes, I think you're right that the issue is with MueLu's ETI logic.

**Comment by jhux2** [comment]
@bartlettroscoe Here's a third, better option.

`-DMueLu_ENABLE_Epetra=OFF`

It has the advantage of not doing unnecessary instantiation at the Tpetra level

**Comment by mhoemmen** [comment]
@bartlettroscoe I don't want to hinder this process; I just want developers to be aware that enabling stuff that most users don't need or (shouldn't) use, might be easy to do, but increases build times and sizes.  I don't want developers to get complacent about that.  If we decide for now not to fix it, that's fine, but that needs to be a conscious choice.

**Comment by mhoemmen** [comment]
In summary: Please go ahead and do what you need to do, but be aware that we're building more than we need.

**Comment by bartlettroscoe** [comment]
> In summary: Please go ahead and do what you need to do, but be aware that we're building more than we need.

I would rather build less.  But given the option of having a build fail with link failures or building more than the user really needs (but building what they are actually asking for), I think we should error on the side of having the build succeeded.

Could problems like this be avoided if MueLu and other packages were better broken up into subpackages?  For example, if Panzer only needs the Tpetra adapters from MueLu, and if MeuLu was broken up into subpackages MueLuCore, MueLuEpetra and MueLuTpetra, then if Panzer only defined a dependency on MueLuCore and MueLuTpetra, then the Epetra adapters would never get built and this problem would not exist.

**Comment by tawiesn** [comment]
> Could problems like this be avoided if MueLu and other packages were better broken up into subpackages?

No, this would not help. At least not for MueLu, since we already have a place for a clean distinction of the Epetra and Tpetra specific code with the Xpetra package. In contrast, it would break one of the core philosophies of MueLu: to be independent of the underlying linear algebra (either Epetra or Tpetra (+ Kokkos as option)). The problem is that people always start using Tpetra in MueLu directly (instead of using Xpetra). Xpetra is meant to deal with the guards and correct instantiations. The problem with Xpetra is, that it needs some more work to enable (or write stubs) for the fancy new features in Tpetra and not all developers are willing or able to invest that additional time and effort doing so. But it's crucial to understand that we do not want to break MueLu apart into two independent pieces (the Epetra and Tpetra part). MueLu provides implementations for rather general multigrid algorithms independent of Epetra and Tpetra.

**Comment by tawiesn** [comment]
@bartlettroscoe @jhux2 
For example: i just found that comment in the source code of the ProjectorSmootherFactory:

> // TAW: Oct 16 2015: subCopy is not part of Xpetra. One should either add it to Xpetra 
> //                   or replace this call by a local loop. I'm not motivated to do this now...

We only misuse Tpetra directly (instead of Xpetra) since in Xpetra we have no routine "subCopy", yet (which is available in Tpetra but not Epetra and therefore also not in Xpetra). The right solution would be to either avoid that function call (by doing it locally by hand) or add that functionality to Xpetra. Then we would not have such linker problems. It seems that this code has not been touched for more than two years. Obviously nobody was interested in that feature too much. Maybe one should just delete these algorithms or move them into a separate optional subpackage of all non-maintained code.

**Comment by bartlettroscoe** [comment]
I pushed the commit eee871d803e2d0a60c60710071f920365687fdcb which sets `MueLu_ENABLE_Epertra=OFF` instead of `Tpetra_INST_SERIAL=ON` and it passed all of the builds, including all of the CUDA builds.  This also cleared up all of the failing MueLu tests (except for one build and that looks to be a separate issue).

Also, using `MueLu_ENABLE_Epertra=OFF` instead of `Tpetra_INST_SERIAL=ON` cut the cumulative package-by-package build time from 3h9m43s to 2h53m65s so the savings is not insignificant.

@bathmatt, it is acceptable for EMPIRE if `MueLu_ENABLE_Epertra=OFF` is set?  Does EMPIRE need Epetra support under MueLu?

<details>
<summary>
  <b>DETAILS</b> (click to expand)
</summary>

Last night I pushed the commit eee871d803e2d0a60c60710071f920365687fdcb:

```
commit eee871d803e2d0a60c60710071f920365687fdcb
Author: Roscoe A. Bartlett <rabartl@sandia.gov>
Date:   Wed Mar 14 17:34:01 2018 -0600

    Set MueLu_ENABLE_Epertra=OFF to fix MueLu CUDA link failures (#2319, TRIL-171)
    
    This should fix the MueLu build failures with CUDA reported in #2319.
    
    I also removed setting Tpetra_INST_SERIAL=ON.  This should cut down on the
    build times over setting Tpetra_INST_SERIAL=ON.

diff --git a/cmake/std/atdm/ATDMDevEnv.cmake b/cmake/std/atdm/ATDMDevEnv.cmake
index 39a8390..53ac0d8 100644
--- a/cmake/std/atdm/ATDMDevEnv.cmake
+++ b/cmake/std/atdm/ATDMDevEnv.cmake
@@ -123,12 +123,10 @@ ATDM_SET_CACHE(Kokkos_ENABLE_Debug_Bounds_Check "${ATDM_BOUNDS_CHECK}" CACHE BOO
 ATDM_SET_CACHE(KOKKOS_ARCH "$ENV{ATDM_CONFIG_KOKKOS_ARCH}" CACHE STRING)
 ATDM_SET_CACHE(EpetraExt_ENABLE_HDF5 OFF CACHE BOOL)
 ATDM_SET_CACHE(MueLu_ENABLE_Experimental ON CACHE BOOL)
+ATDM_SET_CACHE(MueLu_ENABLE_Epetra OFF CACHE BOOL)
 ATDM_SET_CACHE(Panzer_ENABLE_FADTYPE "Sacado::Fad::DFad<RealType>" CACHE STRING)
 ATDM_SET_CACHE(Phalanx_KOKKOS_DEVICE_TYPE "${ATDM_NODE_TYPE}" CACHE STRING)
 ATDM_SET_CACHE(Phalanx_SHOW_DEPRECATED_WARNINGS OFF CACHE BOOL)
-IF (ATDM_USE_CUDA)
-  ATDM_SET_CACHE(Tpetra_INST_SERIAL "${ATDM_USE_CUDA}" CACHE BOOL)
-ENDIF()
 ATDM_SET_CACHE(Tpetra_INST_CUDA "${ATDM_USE_CUDA}" CACHE BOOL)
 ATDM_SET_CACHE(Xpetra_ENABLE_Experimental ON CACHE BOOL)
```

This resulted in all passing builds for MueLu on all of the platforms, including all of the CUDA builds as shown at:

* https://testing-vm.sandia.gov/cdash/index.php?project=Trilinos&date=2018-03-15&filtercombine=and&filtercount=2&showfilters=1&filtercombine=and&field1=buildname&compare1=63&value1=-atdm-&field2=subprojects&compare2=93&value2=MueLu

The only test failures were for the build `Trilinos-atdm-white-ride-cuda-opt` on `white` and that looks to be assoicated with how the tests are being run and not a MueLu problem at this point.  (I will create another GitHub issue to look into that problem.)

It is also interesting to see the impact this has on the build times for the MueLu build wtih CUDA.  Looking at the build ??? on shiller over the last week at:

* https://testing-vm.sandia.gov/cdash/index.php?project=Trilinos&date=2018-03-15&filtercombine=and&filtercount=3&showfilters=1&filtercombine=and&field1=site&compare1=63&value1=hansen&field2=buildname&compare2=61&value2=Trilinos-atdm-hansen-shiller-cuda-opt&field3=buildstarttime&compare3=84&value3=now

The build two days ago on 2018-03-13 was failing and it took 2h50m17s.  Then yesterday, the build passed using `Tpetra_INST_SERIAL=ON` and it took 3h9m43s.  And then today using `MueLu_ENABLE_Epetra=OFF` (and `Tpetra_INST_SERIAL=OFF` impicitly set), it took 2h53m65s.  Therefore, we can see the approach of using `MueLu_ENABLE_Epetra=OFF` vs. `Tpetra_INST_SERIAL=ON` looks to have shaved off about 16m out of a build that takes about 3 hours (using a package-by-package build so build times are a bit inflated).  But that is not too bad.  And you can't really compare the build time to the case where the build failed because there was a bunchy of executbles that aborted their link because they were missing link symbols.

</details>

**Comment by bartlettroscoe** [comment]
I talked with @bathmatt yesterday and he confirmed that EMPIRE does not need Epetra support under MueLu.  Therefore, this is resolved and I am closing this as completed.
